onnxruntime/contrib_ops/cuda/bert/attention_softmax.h

```diff
@@ -292,16 +291,11 @@ __global__ void SoftmaxWithRawMaskLargeKernel(const int all_sequence_length,
       }
     }
 
-    int mask_offset = 0;
     const int batch_index = blockIdx.y;
-    if (mask_dimension == 2) {
-      mask_offset = batch_index * all_sequence_length + seq_idx;
-    } else if (mask_dimension == 3) {
-      mask_offset = (batch_index * sequence_length + sequence_index) * all_sequence_length + seq_idx;
-    } else if (mask_dimension == 4) {
-      int from_index = all_sequence_length - sequence_length + sequence_index;
-      mask_offset = (batch_index * max_sequence_length + from_index) * max_sequence_length + seq_idx;
-    }
+    int mask_offset = attention_mask_strides.x * batch_index +
+                      attention_mask_strides.y * 0 +
```

Comment:
This line will always be zero, why do we add it?

Reply:
Just to make the logic more clear. It will be just optimized away in this case.

TLDR is, we support 2D, 3D, 4D mask. However, for the 4D mask, it is of shape Bx1xSxT. So the third (fast) dimension is useless here. A step back view to this problem is, the attention score will always be 3D (BxSxT). For 2D, it is [Bx1xT] and is broadcasted along the second axis. For a 4D mask, it is actually a 3D mask with different **viewing pattern**. The indexing pattern is always the same and is molded into the kernel impl. The stride is then the **viewing pattern**. So, for an even better/further refactor, we only want to support 3D mask. We then create a `enum` to classify the mask type, then create mask stride by the mask type (with a helper function outside of our kernel).

But if you'd prefer the `enum` and 3D mask. I think I can impl it. We better update the schema if possible, and docs.

Reply:
4D mask is Megatron style which has shape B x 1 x M x M, where M is a fixed value (max sequence length).

Reply:
Let't hold on this PR. I will do some further change in rocm branch first